mars/services/storage/transfer.py

```diff
@@ -214,31 +247,49 @@ async def do_write(self, message: TransferMessage):
         # close may be a high cost operation, use create_task
         close_tasks = []
         finished_keys = []
+        session_id = message.session_id
         for data, data_key, is_eof in zip(message.data,
                                           message.data_keys,
                                           message.eof_marks):
             writer, _, _ = self._key_to_writer_info[
-                (message.session_id, data_key)]
+                (session_id, data_key)]
             if data:
                 await writer.write(data)
             if is_eof:
                 close_tasks.append(asyncio.create_task(writer.close()))
                 finished_keys.append(data_key)
         await asyncio.gather(*close_tasks)
         for data_key in finished_keys:
-            self._key_to_writer_info.pop((message.session_id, data_key))
+            self._key_to_writer_info.pop((session_id, data_key))
+            self._writing_keys[(message.session_id, data_key)].is_success = True
+            self._writing_keys[(session_id, data_key)].set()
+            self._decref_writing_key(session_id, data_key)
 
     async def receive_part_data(self, message: TransferMessage):
+        write_task = asyncio.create_task(self.do_write(message))
         try:
-            yield self.do_write(message)
+            await asyncio.shield(write_task)
         except asyncio.CancelledError:
+            session_id = message.session_id
             for data_key in message.data_keys:
                 if (message.session_id, data_key) in self._key_to_writer_info:
-                    writer, data_size, level = self._key_to_writer_info[
-                        (message.session_id, data_key)]
-                    await self._quota_refs[level].release_quota(data_size)
-                    await self._storage_handler.delete(
-                        message.session_id, data_key, error='ignore')
-                    await writer.clean_up()
-                    self._key_to_writer_info.pop((
-                        message.session_id, data_key))
+                    if self._writing_refs[(message.session_id, data_key)] == 1:
+                        writer, data_size, level = self._key_to_writer_info[
```

Comment:
When deciding to cancel certain transfer task, it shall be popped from `_key_to_writer_info` immediately, or there might be race conditions.